hyperledger/sawtooth-core
# Enforce and abide by PoET key block claim limit

Introduces the following changes:

* PoET consensus state statistics are now reset whenever a non-PoET block is encountered in the chain and the sealed signup data is removed from the consensus state as that needs to be maintained on a per-PoET public key basis.
* A utility class to manage and persist locally on a validator key state for its PoET key.  This includes the sealed signup data associated with it as well as whether or not it has been refreshed because the block claim limit has been reached with that key.
* The PoET genesis CLI and block publisher have been changed to use the PoET key state store for sealed signup data.
* A utility class to query PoET configuration settings.
* The PoET block verifier now enforces the key block claim limit when determining the validity of a block.
* The PoET block publisher abides by the key block claim limit when being asked to initialize a block.  If the limit has been reached with its current key, it will fail to initialize the block and if appropriate will create and register new signup information.

To test locally:
 `cd /project/sawtooth-core/integration/sawtooth_integration/docker` 

In `poet-smoke.yaml`, after the line:
`-k /etc/sawtooth/keys/validator.wif \`

Add:

```
sawtooth.consensus.algorithm=poet \
sawtooth.poet.target_wait_time=5 \
sawtooth.poet.initial_wait_time=0 \
sawtooth.poet.key_block_claim_limit=2 \
```

This should result in a bunch of blocks and validators having to refresh after successfully claiming 2 blocks.

Execute:
`run_docker_test poet-smoke.yaml`

Go get your beverage of choice and cross your fingers that the test passes.
